particle-iot/device-os
# Fixes Boron charging current limitation when powered by VUSB

### Problem

Customers are limited to 500mA charging current when Boron is powered by VUSB.

Affected Device: Gen3 devices

### Analysis
When Boron is powered by VUSB pin, the power supply is blocked by the diode. For nRF52840,
> The USB peripheral has a dedicated internal voltage regulator for converting the VBUS supply to 3.3 V used by the USB signalling interface (D+ and D- lines, and pull-up on D+). The rest of the USB peripheral (USBD) is supplied through the main supply like any other on-chip feature. As a consequence, both VBUS and either VDDH or VDD supplies are required for USB peripheral operation.

![image](https://user-images.githubusercontent.com/7424522/60857339-c3e7da00-a23c-11e9-9bbf-8e2450f051d5.png)

Since the USB peripheral is left to an uncontrolled state, PMIC would wrongly assume the board is connected to a USB host.

### Solution

1. Add a workaround in system power manager to manually increase charging current limit and decrease input voltage limit.
2. Add an API to detect USB state

When the PMIC driver detects the USB host connection, it will also check whether USB is really connected. If not, it will increase charging current.

### Steps to Test

1. Charge Boron via VUSB pin, observe whether the charging current is more than 500mA, should be around 800mA.
2. Connect Boron to a USB host, observe whether the charging current is no more than 500mA.

### Test result

![image](https://user-images.githubusercontent.com/7424522/60857772-59379e00-a23e-11e9-81ef-107df8ba36ed.png)

![image](https://user-images.githubusercontent.com/7424522/60857780-5f2d7f00-a23e-11e9-92f9-0e64c51cb7d2.png)


### Example App

N/A

### References

[CH34730]

---

### Completeness

- [enhancement] [Boron/BSoM] power: Increases charging current to 900mA when powered through VIN (VUSB) pin
